halo2/src/recursion/endoscale.rs

```diff
@@ -0,0 +1,102 @@
+//! Gadget for endoscaling.
+
+use ff::PrimeFieldBits;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter},
+    plonk::Error,
+};
+use pasta_curves::arithmetic::CurveAffine;
+
+mod chip;
+pub mod primitive;
+
+/// Instructions to map bitstrings to and from endoscalars.
+pub trait EndoscaleInstructions<C: CurveAffine>
+where
+    C::Base: PrimeFieldBits,
+{
+    /// A bitstring to be used in endoscaling.
+    type Bitstring: std::fmt::Debug + Clone;
+    /// The maximum number of bits that can be represented by [`Self::Bitstring`].
+    const MAX_BITSTRING_LENGTH: usize;
+
+    /// Witnesses a slice of bools as a vector of [`Self::Bitstring`]s.
+    fn witness_bitstring(bits: &[bool]) -> Vec<Self::Bitstring>;
+
+    /// Computes commitment (Alg 1) to a variable-length bitstring using the endoscaling
+    /// algorithm. Uses a fixed base.
+    ///
+    /// The bitstring is decomposed into pairs of bits using a running sum outside of
+    /// this gadget.
+    ///
+    /// # Panics
+    /// Panics if `NUM_BITS` is larger than `MAX_BITSTRING_LENGTH`.
+    /// Panics if `NUM_BITS` is an odd number.
+    #[allow(clippy::type_complexity)]
+    fn endoscale_fixed_base<L: Layouter<C::Base>, const NUM_BITS: usize, const NUM_WINDOWS: usize>(
+        &self,
+        layouter: L,
+        base: C,
+        bitstring: &Self::Bitstring,
+    ) -> Result<
+        (
+            AssignedCell<C::Base, C::Base>,
+            AssignedCell<C::Base, C::Base>,
+        ),
```

Comment:
Right now we assume that the endoscaling algorithm doesn't return the point at infinity, but in theory there are some solutions for the bitstring that would cause this to happen. In the past we've dismissed this as unlikely because the endoscalars were just random challenges, but in other situations where the value is partially or fully controlled by an adversary we could have problems. Let's revisit this with @daira.

Reply:
Search for "We have glossed over a complication." in the [Halo paper](https://eprint.iacr.org/2019/1021.pdf): we proved that incomplete addition can safely be used for a *single* encoscaling.

Note, however, that this argument does not apply when we transpose the loops over i and j in a multi-endoscaling as done here: https://github.com/zcash/halo2/pull/599/files#diff-a884e132e768c2b2d425500ae7976ae5669e9a0cd135159ece1217209a4ec68aR57-R69

If you want I can try to prove that finding an exceptional case in that algorithm entails finding a discrete log relation between the bases. That's plausible, but it will be some effort to do the proof, so before that I want to make sure that's the algorithm that we're actually going to be using.

We should also estimate the circuit cost of a plain multiscalar mul (i.e. not endoscaling), to check whether endoscaling still makes sense &mdash; because the rationale for including it when we were the Halo paper was partly based on the relative cost in Sonic-style arithmetization. (I feel bad that @therealyingtong has already done lots of work on endoscaling without us having done that already, but we were busy with NU5.)